### PR TITLE
TGP-539 Create NIRMS Number Page

### PR DESCRIPTION
**Pull request by vinoocs**

### CODE REVIEW

File: app/forms/NirmsNumberFormProvider.scala [hunk 1]
```diff
@@ -14,14 +14,19 @@
...
 
-class FakeNavigator(desiredRoute: Call) extends Navigator {
+import javax.inject.Inject
+
+class NirmsNumberFormProvider @Inject() extends Mappings {
+
+  def apply(): Form[String] =
+    Form(
+      "nirmsNumber" -> text("nirmsNumber.error.required")
+        .verifying(regexp("RMS-(GB|NI)-[0-9]{6}", "nirmsNumber.error.invalidFormat"))
```
Comment (liandra-giacomine) [inline app/forms/NirmsNumberFormProvider.scala]: I believe the dashes should be optional

File: app/viewmodels/govuk/LabelFluency.scala [hunk 2]
```diff
@@ -40,6 +40,11 @@ trait LabelFluency {
     def withCssClass(className: String): Label =
       label.copy(classes = s"${label.classes} $className")
 
+    def asHidden(): Label =
```
Comment (liandra-giacomine) [inline app/viewmodels/govuk/LabelFluency.scala]: Doesn't look like this is used anywhere

File: app/views/NirmsNumberView.scala.html [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        @if(form.errors.nonEmpty) {
+            @govukErrorSummary(ErrorSummaryViewModel(form))
+        }
+
+        @govukInput(
+            InputViewModel(
+                field = form("nirmsNumber"),
+                label = LabelViewModel(messages("nirmsNumber.heading")).asPageHeading(Large)
+            ).withHint(Hint(content = Text(messages("nirmsNumber.hint"))))
+            .withWidth(Fixed10)
```
Comment (liandra-giacomine) [inline app/views/NirmsNumberView.scala.html:51]: Does this limit the number of characters?

File: it/test/controllers/NirmsNumberControllerISpec.scala [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+      val result = callRoute(FakeRequest(routes.NirmsNumberController.onSubmit).withFormUrlEncodedBody("nirmsNumber" -> "RMS-GB-123456"))
+
+      status(result) mustBe SEE_OTHER
+
+      redirectLocation(result) mustBe Some(routes.DummyController.onPageLoad.url)
+
+    }
+
+    "return bad request when submitting invalid data" in {
+
```
Comment (liandra-giacomine) [inline it/test/controllers/NirmsNumberControllerISpec.scala:83]: This test and the one below seem to check for the same thing, what is the difference? Should it be testing for a different error string/scenario?

File: project/AppDependencies.scala [hunk 5]
```diff
@@ -9,14 +9,16 @@ object AppDependencies {
     play.sbt.PlayImport.ws,
     "uk.gov.hmrc"       %% "play-frontend-hmrc-play-30" % "8.5.0",
     "uk.gov.hmrc"       %% "bootstrap-frontend-play-30" % bootstrapVersion,
-    "uk.gov.hmrc.mongo" %% "hmrc-mongo-play-30"         % hmrcMongoVersion
+    "uk.gov.hmrc.mongo" %% "hmrc-mongo-play-30"         % hmrcMongoVersion,
```
Comment (liandra-giacomine) [inline project/AppDependencies.scala]: don't need mockito here as it's a test library so only needed in the libraries for test below

File: test/controllers/NirmsNumberControllerSpec.scala [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+      val formWithErrors = formProvider().bind(Map("nirmsNumber" -> inCorrectnirmsNumber))
+
+      val fakeRequestWithData = FakeRequest().withFormUrlEncodedBody("nirmsNumber" -> inCorrectnirmsNumber)
+
+      val result = nirmsNumberController.onSubmit(fakeRequestWithData)
+
+      status(result) mustEqual BAD_REQUEST
+
+      contentAsString(result) mustEqual nirmsNumberView(formWithErrors)(fakeRequest, stubMessages()).toString
```
Comment (liandra-giacomine) [inline test/controllers/NirmsNumberControllerSpec.scala]: I think we decided as a group to test the specific error scenarios in the unit tests for form failures, can we check that the error message returned for a form error is present in these e.g.
`contentAsString(result) mustContain "the error message we return"`

File: test/forms/NirmsNumberFormProviderSpec.scala [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+import play.api.data.{Form, FormError}
+
+class NirmsNumberFormProviderSpec extends StringFieldBehaviours {
+  private val formProvider       = new NirmsNumberFormProvider()
+  private val form: Form[String] = formProvider()
+  private val requiredKey        = "nirmsNumber.error.required"
+  private val invalidKey         = "nirmsNumber.error.invalidFormat"
+  private val rmsRegex           = "RMS-(GB|NI)-[0-9]{6}"
+
+  val nirmsNumberGenerator: Gen[String] = {
```
Comment (liandra-giacomine) [inline test/forms/NirmsNumberFormProviderSpec.scala:30]: Love the usage of Generators here! <3

File: app/views/NirmsNumberView.scala.html (hunk 3 above)
Comment (vinoocs) [reply app/views/NirmsNumberView.scala.html:51]: It just reduces the size of the text field, inline with the prototype. Does not limit the number of characters :)

File: app/forms/NirmsNumberFormProvider.scala (hunk 1 above)
Comment (vinoocs) [reply app/forms/NirmsNumberFormProvider.scala]: Ah, will confirm with the design team then? The example looks a bit confusing in that case.

File: it/test/controllers/NirmsNumberControllerISpec.scala (hunk 4 above)
Comment (vinoocs) [reply it/test/controllers/NirmsNumberControllerISpec.scala:83]: Just another testcase :) It's all sorted now

File: project/AppDependencies.scala (hunk 5 above)
Comment (vinoocs) [reply project/AppDependencies.scala]: Removed it :)

**Review by yohan-oc — DISMISSED** [review]

File: test/controllers/NirmsNumberControllerSpec.scala (hunk 6 above)
Comment (vinoocs) [reply test/controllers/NirmsNumberControllerSpec.scala]: Added the check for the message.

**Review by mycp98 — DISMISSED** [review]
looks good to me 🙂

File: conf/messages.en [hunk 8]
```diff
@@ -97,8 +112,5 @@ signedOut.title = For your security, we signed you out
 signedOut.heading = For your security, we signed you out
 signedOut.guidance = We did not save your answers.
 
-unauthorised.title = There is a problem
```
Comment (besscerule) [inline conf/messages.en]: don't delete these!!!

File: it/test/helpers/ItTestBase.scala [hunk 9]
```diff
@@ -79,8 +79,8 @@ trait ItTestBase extends PlaySpec with GuiceOneServerPerSuite {
   def redirectUrl(response: WSResponse): Option[String] =
     response.header(HeaderNames.LOCATION)
   def callRoute[A](fakeRequest: FakeRequest[A], requiresAuth: Boolean = true)(implicit
-    app: Application,
-    w: Writeable[A]
+                                                                              app: Application,
```
Comment (besscerule) [inline it/test/helpers/ItTestBase.scala]: this feels like bad formating

**Review by besscerule — CHANGES_REQUESTED** [review]

File: it/test/helpers/ItTestBase.scala (hunk 9 above)
Comment (yohan-oc) [reply it/test/helpers/ItTestBase.scala]: This is fixed

File: conf/messages.en (hunk 8 above)
Comment (yohan-oc) [reply conf/messages.en]: This is a bad mistake. This is fixed now

Comment (besscerule) [reply conf/messages.en]: has this been fixed?

Comment (yohan-oc) [reply conf/messages.en]: yes. https://github.com/hmrc/trader-goods-profiles-frontend/blob/TGP-539/create-nirms-number-page/conf/messages.en

**Review by besscerule — DISMISSED** [review]

**Review by besscerule — DISMISSED** [review]

**Review by Ewan-James-Donovan — DISMISSED** [review]
LGTM

**Review by Ewan-James-Donovan — APPROVED** [review]

**Review by mycp98 — APPROVED** [review]

**Pull request merged**